sw/host/opentitanlib/src/spiflash/sfdp.rs

```diff
@@ -0,0 +1,529 @@
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("IO error: {0:?}")]
+    Io(std::io::Error),
+    #[error("the range {0}..{1} is out of bounds")]
+    SliceRange(usize, usize),
+}
+
+#[derive(Debug, Serialize)]
+/// The SFDP header identifies a valid SFDP, its version and the number of parameter headers.
+pub struct SfdpHeader {
+    pub signature: u32,
+    pub minor: u8,
+    pub major: u8,
+    pub nph: u8,
+    pub reserved: u8,
+}
+
+impl TryFrom<&[u8]> for SfdpHeader {
+    type Error = Error;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let mut rdr = std::io::Cursor::new(buf);
+        Ok(SfdpHeader {
+            signature: rdr.read_u32::<LittleEndian>().map_err(|e| Error::Io(e))?,
+            minor: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            major: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            nph: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            reserved: rdr.read_u8().map_err(|e| Error::Io(e))?,
+        })
+    }
+}
+
+#[derive(Debug, Serialize)]
+/// The SFDP parameter header identifies additional parameter tables within the SFDP.
+/// All devices are required to have a JEDEC parameter header and corresponding table.
+pub struct SfdpPhdr {
+    pub id: u8,
+    pub minor: u8,
+    pub major: u8,
+    pub dwords: u8,
+    pub offset: u32,
+}
+
+impl TryFrom<&[u8]> for SfdpPhdr {
+    type Error = Error;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let mut rdr = std::io::Cursor::new(buf);
+        Ok(SfdpPhdr {
+            id: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            minor: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            major: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            dwords: rdr.read_u8().map_err(|e| Error::Io(e))?,
+            offset: rdr.read_u32::<LittleEndian>().map_err(|e| Error::Io(e))? & 0x00FFFFFFu32,
+        })
+    }
+}
+
+// The internal JEDEC parameter struct is used as an intermediate representation
+// in creating the user-visible JEDEC parameter struct.
+struct InternalJedecParams {
+    pub data: Vec<u32>,
+}
+
+impl TryFrom<&[u8]> for InternalJedecParams {
+    type Error = Error;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let mut rdr = std::io::Cursor::new(buf);
+        let mut data = vec![0u32; buf.len() / 4];
+        rdr.read_u32_into::<LittleEndian>(&mut data)
+            .map_err(|e| Error::Io(e))?;
+        Ok(InternalJedecParams { data })
+    }
+}
+
+// The JEDEC parameter table is documented in JESD216 "Serial Flash Discoverable Parameters".
+// The format of the parameter table is expressed in that document as a set of bitfields
+// in the various "dwords" which make up the table.
+//
+// We use the `field` macro to express how to extract and convert the various bitfields
+// into useful values.
+macro_rules! field {
+    // Extract to bool.
+    ($name:ident -> bool, $field:expr, $bitoffset:expr, $bitsize:expr) => {
+        pub fn $name(&self) -> bool {
```

Comment:
Can I make you generate comments for these? Consider using the `paste` crate if so...

Reply:
I'm not sure what you're asking for.  I've added some additional documentation on the workings of the generated functions.

If you're asking me to auto-doc the generated functions, I'd retort with "they're never meant to be seen outside this module".  The generated functions live only in the non-public struct `InternalJedecParams` and are used in constructing the publicly visible `JedecParams` struct which documents all of its fields.